src/amoc_config/amoc_config_attributes.erl

```diff
@@ -0,0 +1,236 @@
+%%==============================================================================
+%% Copyright 2020 Erlang Solutions Ltd.
+%% Licensed under the Apache License, Version 2.0 (see LICENSE file)
+%%==============================================================================
+%% this module is responsible for parsing module attributes
+%%==============================================================================
+-module(amoc_config_attributes).
+
+-include("amoc_config.hrl").
+
+%% API
+-export([get_module_configuration/3]).
+-export([none/1, none/2]).
+-ifdef(TEST).
+-export([%% exported for testing only
+         get_module_attributes/2,
+         process_module_attributes/3]).
+-endif.
+
+-type maybe_module_attribute() :: module_attribute() | term().
+-type maybe_verification_method() :: verification_method() | term().
+-type maybe_update_method() :: update_method() | term().
```

Comment:
These types are absolutely correctly representing the things that can be supplied. You cannot verify attributes types using dialyzer, but we want for sure at least runtime verification!  BTW without these types dialyzer complains that some of the case-of clauses are impossible to reach!

Reply:
First of all, I think there is no need to shout. Maybe that's just me but when I'm reading a sentence finished with `!` I think the person writing it is shouting at me, I don't like that.

Other than that, I can see that only the `get_module_configuration` function is used from other modules. Other functions seems to be internal or used in tests. 
Do you think it would be possible to modify the `get_module_attributes` function to return data of type `[module_attribute]`? It would require some filtering but if done in this place all the other function don't have to expect unexpected types. It looks like, thanks to that change we could reduce number of function expecting `maybe_*` type significantly.

Moreover, the `maybe` type suggests (from a theoretical perspective) that it contains either some real data or `none` and not anything as it is in this module.

Reply:
`module_attribute()` defines the only correct way of specifying the attribute, we cannot use it until we finish verification of the functions. We cannot be sure that it's a correct `module_attribute()` type. But I think we can make 'maybe_module_attribute()' type spec more restrictive.

> a sentence finished with ! I think the person writing it is shouting at me, I don't like that.

hm, I always thought that [ALL CAPS](https://en.wikipedia.org/wiki/All_caps) is the way to express shouting in print :)

Reply:
Yeah, all caps is an obvious way of shouting in print. The exclamation mark often expresses excitement but shouting as well.